File: infra/deploy/verify-shared-edge-receipt.mjs
#!/usr/bin/env node
import { isIP } from 'node:net';
import { readFileSync } from 'node:fs';

function fail(message) {
  process.stderr.write(`shared edge receipt rejected: ${message}\n`);
  process.exit(1);
}

function exactKeys(value, expected, label) {
  if (!value || typeof value !== 'object' || Array.isArray(value)) fail(`${label}_shape`);
  const actual = Object.keys(value).sort();
  const wanted = [...expected].sort();
  if (JSON.stringify(actual) !== JSON.stringify(wanted)) fail(`${label}_shape`);
}

function validSha(value) {
  return typeof value === 'string' && /^[a-f0-9]{40}$/.test(value);
}

const [
  receiptPath,
  expectedStatus,
  ownerProject,
  ownerRepo,
  operationSha,
  runtimeSha,
  workflowRunRaw,
  phase,
  rollbackFromSha = '',
  rollbackTargetSha = '',
] = process.argv.slice(2);

if (![8, 10].includes(process.argv.slice(2).length)) fail('usage');
if (!['success', 'failure'].includes(expectedStatus)) fail('expected_status');
if (!/^[a-z][a-z0-9_-]{1,31}$/.test(ownerProject ?? '')) fail('owner_project');
if (!/^[A-Za-z0-9_.-]+\/[A-Za-z0-9_.-]+$/.test(ownerRepo ?? '')) fail('owner_repo');
if (!validSha(operationSha) || !validSha(runtimeSha)) fail('sha');
if (!/^[1-9][0-9]*$/.test(workflowRunRaw ?? '')) fail('workflow_run');
if (!['pre-mutation', 'pre-activation', 'post-activation', 'post-rollback', 'post-compensation'].includes(phase)) fail('phase');
const rollbackExpected = rollbackFromSha !== '' || rollbackTargetSha !== '';
if (rollbackExpected && ![rollbackFromSha, rollbackTargetSha].every(validSha)) fail('rollback_metadata');
if ((phase === 'post-rollback' || phase === 'post-compensation') !== rollbackExpected) fail('rollback_phase');
if (phase === 'post-activation' && runtimeSha !== operationSha) fail('post_activation_runtime');
if (phase === 'post-rollback' && (rollbackFromSha === rollbackTargetSha || runtimeSha !== rollbackTargetSha)) fail('post_rollback_runtime');
if (phase === 'post-compensation' && (rollbackFromSha === rollbackTargetSha || runtimeSha !== rollbackFromSha)) fail('post_compensation_runtime');

let receipt;
try {
  receipt = JSON.parse(readFileSync(receiptPath, 'utf8'));
} catch {
  fail('unreadable');
}

exactKeys(receipt, ['contractVersion', 'owner', 'operation', 'workflowRun', 'runtime', 'rollback', 'phase', 'timestamp', 'overallStatus', 'urls', 'edge'], 'receipt');
exactKeys(receipt.owner, ['project', 'repo'], 'owner');
exactKeys(receipt.operation, ['fullSha'], 'operation');
exactKeys(receipt.runtime, ['fullSha'], 'runtime');
exactKeys(receipt.rollback, ['rollbackFrom', 'target'], 'rollback');
if (receipt.contractVersion !== 1) fail('contract_version');
if (receipt.owner.project !== ownerProject || receipt.owner.repo !== ownerRepo) fail('owner_identity');
if (receipt.operation.fullSha !== operationSha || receipt.workflowRun !== Number(workflowRunRaw)) fail('operation_identity');
if (receipt.runtime.fullSha !== runtimeSha || receipt.phase !== phase) fail('runtime_identity');
if (receipt.overallStatus !== expectedStatus) fail('overall_status');
if (typeof receipt.timestamp !== 'string' || !/^\d{4}-\d\d-\d\dT\d\d:\d\d:\d\dZ$/.test(receipt.timestamp)) fail('timestamp');
const expectedRollback = rollbackExpected
  ? { rollbackFrom: rollbackFromSha, target: rollbackTargetSha }
  : { rollbackFrom: null, target: null };
if (JSON.stringify(receipt.rollback) !== JSON.stringify(expectedRollback)) fail('rollback_identity');

const urlContract = new Map([
  ['blog-public', { configuredURL: 'https://blog.blankhoney.xyz/zh', hosts: new Set(['blog.blankhoney.xyz']) }],
  ['blog-public-status', { configuredURL: 'https://blog.blankhoney.xyz/api/status', hosts: new Set(['blog.blankhoney.xyz']) }],
  ['rss-production-auth', { configuredURL: 'https://ai-reader.blankhoney.xyz/', hosts: new Set(['ai-reader.blankhoney.xyz', 'auth.blankhoney.xyz']) }],
]);
if (!Array.isArray(receipt.urls) || receipt.urls.length !== urlContract.size) fail('urls_shape');
const seenNames = new Set();
for (const item of receipt.urls) {
  exactKeys(item, ['name', 'configuredURL', 'status', 'finalURL', 'tls', 'redirect', 'result', 'error'], 'url');
  const contract = urlContract.get(item.name);
  if (!contract || seenNames.has(item.name) || item.configuredURL !== contract.configuredURL) fail('url_identity');
  seenNames.add(item.name);
  if (!['success', 'failure'].includes(item.result)) fail('url_result');
  if (typeof item.tls !== 'boolean' || typeof item.redirect !== 'boolean') fail('url_types');
  if (item.result === 'success') {
    if (item.status !== 200 || item.tls !== true || item.error !== null || typeof item.finalURL !== 'string') fail('url_success');
    let parsed;
    try { parsed = new URL(item.finalURL); } catch { fail('url_final'); }
    if (parsed.protocol !== 'https:' || parsed.username || parsed.password || parsed.port || isIP(parsed.hostname) || !contract.hosts.has(parsed.hostname)) fail('url_final');
    if (item.name.startsWith('blog-public') && item.redirect !== false) fail('blog_redirect');
    if (item.name === 'rss-production-auth' && (item.redirect !== true || parsed.hostname !== 'auth.blankhoney.xyz')) fail('rss_redirect');
  } else {
    const statusValid = item.status === null || (Number.isSafeInteger(item.status) && item.status >= 100 && item.status <= 599);
    const finalValid = item.finalURL === null || typeof item.finalURL === 'string';
    if (!statusValid || !finalValid || item.error === null || !/^[a-z0-9_]+$/.test(item.error)) fail('url_failure');
    if (item.finalURL !== null) {
      let parsed;
      try { parsed = new URL(item.finalURL); } catch { fail('url_failure_final'); }
      if (parsed.protocol !== 'https:' || parsed.username || parsed.password || parsed.port
        || isIP(parsed.hostname) || !contract.hosts.has(parsed.hostname)) fail('url_failure_final');
    }
  }
}
if (seenNames.size !== urlContract.size || [...urlContract.keys()].some((name) => !seenNames.has(name))) fail('urls_missing');

exactKeys(receipt.edge, [
  'caddyContainer', 'myrssAppAttached', 'brianstormEdgeAttached', 'networkDriver',
  'configLoaded', 'rssUpstreamReachable', 'blogUpstreamReachable', 'result', 'error',
], 'edge');
if (receipt.edge.caddyContainer !== 'myrss-edge-caddy-1' || receipt.edge.result !== expectedStatus) fail('edge_identity');
if (expectedStatus === 'success') {
  if (receipt.urls.some((item) => item.result !== 'success')) fail('success_urls');
  if (receipt.edge.error !== null || receipt.edge.networkDriver !== 'bridge'
    || receipt.edge.myrssAppAttached !== true || receipt.edge.brianstormEdgeAttached !== true
    || receipt.edge.configLoaded !== true || receipt.edge.rssUpstreamReachable !== true
    || receipt.edge.blogUpstreamReachable !== true) fail('success_edge');
} else {
  if (!Array.isArray(receipt.edge.error) || receipt.edge.error.length === 0
    || receipt.edge.error.some((value) => typeof value !== 'string' || !/^[a-z0-9_]+$/.test(value))) fail('failure_edge');
}

process.stdout.write(`shared edge ${expectedStatus} receipt verified\n`);

File: infra/deploy/verify-shared-edge-receipt.py
#!/usr/bin/env python3
"""Strict shared-edge receipt verifier using only the Python standard library."""
from __future__ import annotations

import ipaddress
import json
import re
import sys
import urllib.parse
from pathlib import Path

SHA = re.compile(r'^[a-f0-9]{40}$')
RFC3339 = re.compile(r'^\d{4}-\d\d-\d\dT\d\d:\d\d:\d\dZ$')
PHASES = {'pre-mutation', 'pre-activation', 'post-activation', 'post-rollback', 'post-compensation'}
URLS = {
    'blog-public': ('https://blog.blankhoney.xyz/zh', {'blog.blankhoney.xyz'}),
    'blog-public-status': ('https://blog.blankhoney.xyz/api/status', {'blog.blankhoney.xyz'}),
    'rss-production-auth': ('https://ai-reader.blankhoney.xyz/', {'ai-reader.blankhoney.xyz', 'auth.blankhoney.xyz'}),
}


def exact(value: object, keys: set[str], label: str) -> dict:
    if type(value) is not dict or set(value) != keys:
        raise ValueError(f'{label}_shape')
    return value


def safe_url(value: str, hosts: set[str]) -> urllib.parse.SplitResult:
    parsed = urllib.parse.urlsplit(value)
    if parsed.scheme != 'https' or parsed.username or parsed.password or parsed.port is not None:
        raise ValueError('url_transport')
    if parsed.hostname not in hosts:
        raise ValueError('url_host')
    try:
        ipaddress.ip_address(parsed.hostname or '')
    except ValueError:
        return parsed
    raise ValueError('url_ip')


def verify(receipt: object, expected: dict[str, object]) -> None:
    value = exact(receipt, {'contractVersion', 'owner', 'operation', 'workflowRun', 'runtime',
        'rollback', 'phase', 'timestamp', 'overallStatus', 'urls', 'edge'}, 'receipt')
    owner = exact(value['owner'], {'project', 'repo'}, 'owner')
    operation = exact(value['operation'], {'fullSha'}, 'operation')
    runtime = exact(value['runtime'], {'fullSha'}, 'runtime')
    rollback = exact(value['rollback'], {'rollbackFrom', 'target'}, 'rollback')
    if value['contractVersion'] != 1 or owner != {'project': expected['ownerProject'], 'repo': expected['ownerRepo']}:
        raise ValueError('owner_identity')
    if operation['fullSha'] != expected['operationSha'] or value['workflowRun'] != expected['workflowRun']:
        raise ValueError('operation_identity')
    if runtime['fullSha'] != expected['runtimeSha'] or value['phase'] != expected['phase']:
        raise ValueError('runtime_identity')
    if value['overallStatus'] != expected['status'] or not RFC3339.fullmatch(value['timestamp']):
        raise ValueError('status_timestamp')
    if rollback != expected['rollback']:
        raise ValueError('rollback_identity')
    if type(value['urls']) is not list or len(value['urls']) != len(URLS):
        raise ValueError('urls_shape')
    seen = set()
    for item_raw in value['urls']:
        item = exact(item_raw, {'name', 'configuredURL', 'status', 'finalURL', 'tls', 'redirect', 'result', 'error'}, 'url')
        if item['name'] not in URLS or item['name'] in seen:
            raise ValueError('url_identity')
        seen.add(item['name'])
        configured, hosts = URLS[item['name']]
        if item['configuredURL'] != configured or item['result'] not in {'success', 'failure'}:
            raise ValueError('url_contract')
        if item['finalURL'] is not None:
            safe_url(item['finalURL'], hosts)
        if item['result'] == 'success':
            if item['status'] != 200 or item['tls'] is not True or item['error'] is not None:
                raise ValueError('url_success')
            final = safe_url(item['finalURL'], hosts)
            if item['name'].startswith('blog-public') and item['redirect'] is not False:
                raise ValueError('blog_redirect')
            if item['name'] == 'rss-production-auth' and (item['redirect'] is not True or final.hostname != 'auth.blankhoney.xyz'):
                raise ValueError('rss_redirect')
        elif type(item['error']) is not str or not re.fullmatch(r'[a-z0-9_]+', item['error']):
            raise ValueError('url_failure')
        if item['result'] == 'failure':
            if item['status'] is not None and (type(item['status']) is not int or not 100 <= item['status'] <= 599):
                raise ValueError('url_failure_status')
            if item['finalURL'] is not None and type(item['finalURL']) is not str:
                raise ValueError('url_failure_final')
        if type(item['tls']) is not bool or type(item['redirect']) is not bool:
            raise ValueError('url_types')
    if seen != set(URLS):
        raise ValueError('urls_missing')
    edge = exact(value['edge'], {'caddyContainer', 'myrssAppAttached', 'brianstormEdgeAttached',
        'networkDriver', 'configLoaded', 'rssUpstreamReachable', 'blogUpstreamReachable', 'result', 'error'}, 'edge')
    if edge['caddyContainer'] != 'myrss-edge-caddy-1' or edge['result'] != expected['status']:
        raise ValueError('edge_identity')
    if expected['status'] == 'success':
        if edge != {'caddyContainer': 'myrss-edge-caddy-1', 'myrssAppAttached': True,
                'brianstormEdgeAttached': True, 'networkDriver': 'bridge', 'configLoaded': True,
                'rssUpstreamReachable': True, 'blogUpstreamReachable': True,
                'result': 'success', 'error': None}:
            raise ValueError('success_edge')
    elif (type(edge['error']) is not list or not edge['error']
            or any(type(item) is not str or not re.fullmatch(r'[a-z0-9_]+', item) for item in edge['error'])):
        raise ValueError('failure_edge')


def parse_expected(values: list[str]) -> tuple[Path, dict[str, object]]:
    if len(values) not in {8, 10}:
        raise ValueError('usage')
    receipt, status, project, repo, operation, runtime, workflow, phase, *rollback_values = values
    if status not in {'success', 'failure'} or phase not in PHASES or not SHA.fullmatch(operation) or not SHA.fullmatch(runtime):
        raise ValueError('arguments')
    if not re.fullmatch(r'[1-9][0-9]*', workflow):
        raise ValueError('workflow')
    rollback_expected = phase in {'post-rollback', 'post-compensation'}
    if rollback_expected != (len(rollback_values) == 2) or any(not SHA.fullmatch(item) for item in rollback_values):
        raise ValueError('rollback')
    rollback = {'rollbackFrom': rollback_values[0], 'target': rollback_values[1]} if rollback_values else {'rollbackFrom': None, 'target': None}
    if phase == 'post-activation' and runtime != operation:
        raise ValueError('post_activation')
    if phase == 'post-rollback' and (rollback['rollbackFrom'] == rollback['target'] or runtime != rollback['target']):
        raise ValueError('post_rollback')
    if phase == 'post-compensation' and (rollback['rollbackFrom'] == rollback['target'] or runtime != rollback['rollbackFrom']):
        raise ValueError('post_compensation')
    return Path(receipt), {'status': status, 'ownerProject': project, 'ownerRepo': repo,
        'operationSha': operation, 'runtimeSha': runtime, 'workflowRun': int(workflow), 'phase': phase,
        'rollback': rollback}


def main(values: list[str]) -> int:
    receipt_path, expected = parse_expected(values)
    verify(json.loads(receipt_path.read_text()), expected)
    print(f"shared edge {expected['status']} receipt verified")
    return 0


if __name__ == '__main__':
    try:
        raise SystemExit(main(sys.argv[1:]))
    except Exception as error:
        print(f'shared edge receipt rejected: {type(error).__name__}:{error}', file=sys.stderr)
        raise SystemExit(1)
